xdebug_handler_dbgp.c

```diff
@@ -1126,68 +1147,129 @@ DBGP_FUNC(stdout)
 	success = "1";
 
 	xdebug_xml_add_attribute_ex(*retval, "success", xdstrdup(success), 0, 1);
+	return 0;
 }
 
 DBGP_FUNC(stop)
 {
 	XG(status) = DBGP_STATUS_STOPPED;
 	xdebug_xml_add_attribute(*retval, "status", xdebug_dbgp_status_strings[XG(status)]);
 	xdebug_xml_add_attribute(*retval, "reason", xdebug_dbgp_reason_strings[XG(reason)]);
+	return 2;
 }
 
 DBGP_FUNC(run)
 {
-	xdebug_xml_add_attribute_ex(*retval, "filename", xdstrdup(context->program_name), 0, 1);
+	if(XG(status) ==DBGP_STATUS_BREAK||XG(status) ==DBGP_STATUS_STARTING){
+		XG(status) = DBGP_STATUS_RUNNING;
+		XG(reason) = DBGP_REASON_OK;
+		XG(lastcmd) = "run";
+		if (XG(lasttransid)) {
+			xdfree(XG(lasttransid));
+		}
+		XG(lasttransid) = xdstrdup(CMD_OPTION_CHAR('i'));
+
+		xdebug_xml_add_attribute_ex(*retval, "filename", xdstrdup(context->program_name), 0, 1);
+		return 1;
+	} else {
+		RETURN_RESULT(XG(status), XG(reason), XDEBUG_ERROR_COMMAND_UNAVAILABLE);
+	}
 }
 
 DBGP_FUNC(step_into)
 {
-	XG(context).do_next   = 0;
-	XG(context).do_step   = 1;
-	XG(context).do_finish = 0;
+	if(XG(status) ==DBGP_STATUS_BREAK||XG(status) ==DBGP_STATUS_STARTING){
+		XG(status) = DBGP_STATUS_RUNNING;
+		XG(reason) = DBGP_REASON_OK;
+		XG(lastcmd) = "step_into";
+		if (XG(lasttransid)) {
+			xdfree(XG(lasttransid));
+		}
+		XG(lasttransid) = xdstrdup(CMD_OPTION_CHAR('i'));
+
+		XG(context).do_next   = 0;
+		XG(context).do_step   = 1;
+		XG(context).do_finish = 0;
+		return 1;
+	} else {
+		RETURN_RESULT(XG(status), XG(reason), XDEBUG_ERROR_COMMAND_UNAVAILABLE);
+	}
 }
 
 DBGP_FUNC(step_out)
 {
-	function_stack_entry *fse;
+	if(XG(status) == DBGP_STATUS_BREAK||XG(status) ==DBGP_STATUS_STARTING){
+		function_stack_entry *fse;
 
-	XG(context).do_next   = 0;
-	XG(context).do_step   = 0;
-	XG(context).do_finish = 1;
+		XG(status) = DBGP_STATUS_RUNNING;
+		XG(reason) = DBGP_REASON_OK;
+		XG(lastcmd) = "step_out";
+		if (XG(lasttransid)) {
+			xdfree(XG(lasttransid));
+		}
+		XG(lasttransid) = xdstrdup(CMD_OPTION_CHAR('i'));
```

Comment:
There is quite a bit of repetition here. I would do the following:

1. turn the check in line 1201 (and others) into it's own static function
2. Make the check negative, so that you can:
  - use `RETURN_RESULT(XG(status), XG(reason), XDEBUG_ERROR_COMMAND_UNAVAILABLE);` to leave the function immediately
  - not necessarily have to add extra indentation to the bulk of the function(s)
3. create another static helper function to do line 1204-1210